### PR TITLE
refactor-privacy-storage

### PR DESCRIPTION
**Pull request by iikirilov**
Signed-off-by: Ivaylo Kirilov <iikirilov@gmail.com>

## PR description

Refactor of privacy storage. Required for [PAN-3096]
* changed suffix for stored private events for consistency
* added `blockHash_txHash_METADATA` -> `PrivateTransactionMetadata` storage
* ~code to migrate existing privacy database~

### CODE REVIEW

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/PrivateStateKeyValueStorage.java [hunk 1]
```diff
@@ -12,24 +12,34 @@
...
 import org.hyperledger.besu.util.bytes.Bytes32;
 import org.hyperledger.besu.util.bytes.BytesValue;
+import org.hyperledger.besu.util.bytes.BytesValues;
 
+import java.util.List;
 import java.util.Optional;
 
 public class PrivateStateKeyValueStorage implements PrivateStateStorage {
 
+  private static final BytesValue EVENTS_KEY_SUFFIX = BytesValue.of("EVENTS".getBytes(UTF_8));
```
Comment (pinges) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/PrivateStateKeyValueStorage.java]: Do you want to change that to TRANSACTION_LOGS_KEY_SUFFIX to make it consistent?

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/precompiles/privacy/PrivacyPrecompiledContract.java [hunk 2]
```diff
@@ -114,7 +109,7 @@ public BytesValue compute(final BytesValue input, final MessageFrame messageFram
 
     // get the last world state root hash - or create a new one
     final Hash lastRootHash =
-        privateStateStorage.getPrivateAccountState(privacyGroupId).orElse(EMPTY_ROOT_HASH);
+        privateStateStorage.getLatestStateRoot(privacyGroupId).orElse(EMPTY_ROOT_HASH);
```
[the remark this answers is not in the document]
Comment (iikirilov) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/precompiles/privacy/PrivacyPrecompiledContract.java]: This is not related to layer 1 implementation - it is just a method name change to better reflect what the method is doing. This method will be deprecated and a migration procedure will be implemented to change the way the latest state root is accessed.

**Comment by iikirilov** [comment]
> In Besu they are using the concept of segmented storage now. Wouldn't it be better to use that instead of the prefixes?

I believe that is already the case. The private storage is a segment of the db - but the values stored are suffixed similar to `KeyValueStoragePrefixedKeyBlockchainStorage`

**Comment by iikirilov** [comment]
On a side-note looks like the previous comment unassigned @pinges which was not intended. I do not have rights to assign/unassigned people.

File: plugin-api/src/main/java/org/hyperledger/besu/plugin/services/storage/KeyValueStorage.java [hunk 3]
```diff
@@ -70,6 +71,15 @@
...
   long removeAllKeysUnless(Predicate<byte[]> retainCondition) throws StorageException;
 
+  /**
+   * Performs an evaluation against each key in the store, returning the set of entries that pass.
+   *
+   * @param returnCondition predicate to evaluate each key against, unless the result is {@code
+   *     null}, the key is added to the returned list of keys.
+   * @return the set of keys that pass the condition.
+   */
+  Set<byte[]> getAllKeysThat(Predicate<byte[]> returnCondition);
```
Comment (jframe) [inline plugin-api/src/main/java/org/hyperledger/besu/plugin/services/storage/KeyValueStorage.java]: Can't see where this is being used other than in tests. Is it needed?

Comment (jframe) [reply plugin-api/src/main/java/org/hyperledger/besu/plugin/services/storage/KeyValueStorage.java]: I see now that  it's going used in another PR https://github.com/hyperledger/besu/pull/25

Comment (iikirilov) [reply plugin-api/src/main/java/org/hyperledger/besu/plugin/services/storage/KeyValueStorage.java]: this PR is to prepare the rest. The plan was to make it easier for review just to split the storage changes from the rest.

**Pull request merged**